### ISSUE TITLE
Alternator: missing error checks in BatchWriteItem

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#5057]
BatchWriteItem is supposed to limit the number of items (DynamoDB limits to 25) or size of items (400 KB) or total request size (16 MB). We don't have to use the same limits as DynamoDB, but we should probably have limits, and/or make sure we yield while working on a large batch request and that we don't use excessive parallelism (currently our BatchWriteItem implementation does all the updates in parallel).

**Comment by nyh** [comment]
Some insights from the discussion in the duplicate issue #9389:

We already have a test for the BatchWriteItem limit of 25 items, `test_batch.py`::`test_batch_write_item_too_many` (waiting for review on the mailing list, titled "[PATCH] test/alternator: add test for too many items in BatchWriteItem"). It confirms that this limit 25 exists on DynamoDB but not on Scylla.

There is no particular reason why this limit needs to be 25 items specifically. It could be raised to something else (e.g., 50), or be unlimited as it is today (although, please note that there is a limit on the total request size, so it's not completely unlimited). However, there are some consequences to consider:

First, the current implementation of BatchWriteItem executes all writes in the batch in parallel. Because these writes can go to different partitions, this can result in very high parallelism - especially if the user does many BatchWriteItem requests in parallel (we saw one user doing exactly that). So if we don't place a low-enough limit on the number of items in a BatchWriteItem request, we should at least use a semaphore (per batch, or perhaps even a per-shard one?) to apply these writes with fixed parallelism. Adding such a semaphore may be a valid fix for this issue - but we should also document this difference in our documentation.

The second problem is that even if Scylla allows BatchWriteItem requests with 50 or 1000 items, clients have no official mechanism to discover that limit. The only thing they can do is to try some batch size and if it results in an error try again with a smaller batch size. The existing DynamoDB libraries (e.g., the Python boto3) **do not** do that - their recommended API (see for example see boto3's `batch_writer` - https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb.html#DynamoDB.Table.batch_writer) splits batch writes into BatchWriteItem requests of a hard-coded length 25. However, some users might not want to use boto3's high-level `batch_writer` mechanism, and instead want to use the low-level `batch_write_item` function (which generates a `BatchWriteItem` request directly). Those users might appreciate being allowed to put more than 25 items in a request. But they will need to code their own discovery mechanism of what the allowed limit is. So in this regard, the option of having **no limit** (other than the overarching limit on the total request size) is cleaner than the option of enforcing a specific limit such as 50 or 1000.

**Comment by swasik** [comment]
I think that this is a general decision we have to take, what do we prefer:

1. If Scylla has less limitations than Dynamo, do we want to make Alternator API less strict? This way we can provide more value to users but on the other hand we will have to keep this decision for ever, which:
  a. Makes behavior less obvious both for maintainers (Scylla developers) and users.
  b. Even if it works correctly now, in the future can lead to some performance problems.
  c. In some corner cases being incompatible today may make it impossible to be compatible in the future (if Amazon introduces behavior that contradicts our less strict implementation).
2. Keep as much compatibility as possible, which also allows users to freely experiment with Scylla without worries that in the future they will not be able to migrate back to Dynamo. I think that what should distinguish us from Dynamo is performance, cost and support for platforms other than AWS, not some small improvements of API here and there.

Summarizing, I would rather keep the behavior compatible with Dynamo, unless we can provide some big, well understood value by doing the opposite. @mykaul @tzach what do you think?

**Comment by mykaul** [comment]
I think we should have:
1. Configurable values for those limits.
2. Do *NOT* configure them to be DynamoDB compatible - we do want to show we have an advantage here. So for example instead of 400K item size limit we can put 4M as a 10x higher value. 
If anyone wants strict DynamoDB compatibility, they can set it with DynamoDB compatible values.

**Comment by nyh** [comment]
> Summarizing, I would rather keep the behavior compatible with Dynamo, unless we can provide some big, well understood value by doing the opposite. [@mykaul](https://github.com/mykaul) [@tzach](https://github.com/tzach) what do you think?

That is also my personal preference, especially for two reasons:
1. Users would have way to figure out what is the limit (applications won't even try batches of 50 writes because they know this wouldn't work in DynamoDB).
2. If users do use huge patches, the performance will be disappointing (e.g., latency of the entire request or excessive parallelism) and users may complain not realizing they did something which isn't even allowed in DynamoDB. We suspect we had such a problem recently with huge items, which Alternator didn't forbid but then also didn't perform well with them.

But I see @mykaul has the opposite preference. Both preferences have their benefits, I don't think there's an obvious choice here.

**Comment by ScyllaPiotr** [comment]
Maybe we could take an empirical approach here, in the form of the following chronology:
1. Implement a strict DDB-compatibility, which is hardcoded limit of batch sizes up to 25,
2. Experiment with different (higher) batch sizes in the context of a few representational cluster hw/sw configs,
3. Based on the data above, decide whether to implement configurable batch sizes, establish their sensible ranges and default values.

**Comment by mykaul** [comment]
> Maybe we could take an empirical approach here, in the form of the following chronology:
> 
> 1. Implement a strict DDB-compatibility, which is hardcoded limit of batch sizes up to 25,
> 2. Experiment with different (higher) batch sizes in the context of a few representational cluster hw/sw configs,

We have customer(s) running with a size of 100 or so, per batch...

**Comment by ScyllaPiotr** [comment]
> We have customer(s) running with a size of 100 or so, per batch...
So, does this mean that we cannot implement 100% DDB-compatible behaviour (up to 25 batches) any more?

**Comment by swasik** [comment]
Configuration option seems reasonable. And taking into account that we already have users with high batch sizes we should probably set it reasonably high by default (e.g. 200 if we know about scenarios using 100 items)

**Comment by radoslawcybulski** [comment]
Do we have some sort of global configuration object, that is available for quering the limit size? If so, how to reach it?

**Comment by ScyllaPiotr** [comment]
> We have customer(s) running with a size of 100 or so, per batch...

@mykaul we are one step from approving a [PR with a limit of 100 items in a batch](https://github.com/scylladb/scylladb/pull/23232/).
Do you think we should first make sure that the limit of 100 items is acceptable for all our Alternator clients/deployments?
If so, how to check it? Who to ask for data?

**Comment by mykaul** [comment]
> > We have customer(s) running with a size of 100 or so, per batch...
> 
> [@mykaul](https://github.com/mykaul) we are one step from approving a [PR with a limit of 100 items in a batch](https://github.com/scylladb/scylladb/pull/23232/). Do you think we should first make sure that the limit of 100 items is acceptable for all our Alternator clients/deployments? If so, how to check it? Who to ask for data?

@amnonh ?

**Comment by nyh** [comment]
Note that this PR makes this limit of 100 configurable, so if a specific customer needs a higher limit, they can easily set it. But I agree it will be good to know in advance if any existing customer needs a higher limit so they don't encounter problems after upgrade and need to solve it as an emergency instead of preemptively.

Also, importantly, please note that this issue, and the PR, is about **batch writes**. Batch reads (`BatchGetItem`) have a different limit - in DynamoDB that limit is 100, compared to the write limit which is 25. When you think about a customer using a batch of 100 items, maybe you're thinking of reads?

As I mentioned in the past, any application that was ported from DynamoDB must be using DynamoDB's fixed limits (25 for writes, 100 for reads) - because there is no way to discover these limits so the SDK and applications have them hard-coded.

**Comment by amnonh** [comment]
> > > We have customer(s) running with a size of 100 or so, per batch...
> > 
> > 
> > [@mykaul](https://github.com/mykaul) we are one step from approving a [PR with a limit of 100 items in a batch](https://github.com/scylladb/scylladb/pull/23232/). Do you think we should first make sure that the limit of 100 items is acceptable for all our Alternator clients/deployments? If so, how to check it? Who to ask for data?
> 
> [@amnonh](https://github.com/amnonh) ?

Not exactly, we have a counter that counts the total number of items in a batch. We can have an average batch item size but not batch size statistics.

**Comment by ScyllaPiotr** [comment]
> > > > We have customer(s) running with a size of 100 or so, per batch...
> > > 
> > > 
> > > [@mykaul](https://github.com/mykaul) we are one step from approving a [PR with a limit of 100 items in a batch](https://github.com/scylladb/scylladb/pull/23232/). Do you think we should first make sure that the limit of 100 items is acceptable for all our Alternator clients/deployments? If so, how to check it? Who to ask for data?
> > 
> > 
> > [@amnonh](https://github.com/amnonh) ?
> 
> Not exactly, we have a counter that counts the total number of items in a batch. We can have an average batch item size but not batch size statistics.

@amnonh Could you check if anyone anywhere used a write batch containing more than 100 items?

**Comment by mykaul** [comment]
> > > > > We have customer(s) running with a size of 100 or so, per batch...
> > > > 
> > > > 
> > > > [@mykaul](https://github.com/mykaul) we are one step from approving a [PR with a limit of 100 items in a batch](https://github.com/scylladb/scylladb/pull/23232/). Do you think we should first make sure that the limit of 100 items is acceptable for all our Alternator clients/deployments? If so, how to check it? Who to ask for data?
> > > 
> > > 
> > > [@amnonh](https://github.com/amnonh) ?
> > 
> > 
> > Not exactly, we have a counter that counts the total number of items in a batch. We can have an average batch item size but not batch size statistics.
> 
> [@amnonh](https://github.com/amnonh) Could you check if anyone anywhere used a write batch containing more than 100 items?

I'm pretty sure we have had BatchGetItem with >100 items, could not find evidence for Write though.

**Comment by amnonh** [comment]
I can only see average sizes and couldn't find any cluster with batch size greater than 10 (neither get or write)

**Comment by ScyllaPiotr** [comment]
I approved the PR.

**Comment by mykaul** [comment]
> > I'm pretty sure we have had BatchGetItem with >100 items, could not find evidence for Write though.
> 
> BatchGetItem with >100 items, or exactly 100 items? Because >100 would not have worked on DynamoDB, so it's strange that somebody who migrated from DynamoDB would have such a use case.

I remember the mythical number 112, but could not find the evidence right now.

> 
> In any case, this issue is about BatchWriteItem, whose limit on DynamoDB is even lower - 25.

Yes.

**Comment by amnonh** [comment]
@mykaul @nyh What we can do is add a histogram for batch sizes (I wanted to do it anyhow), start with a warning first and we can switch between warning and returning an error by a configuration flag.

**Comment by mykaul** [comment]
> [@mykaul](https://github.com/mykaul) [@nyh](https://github.com/nyh) What we can do is add a histogram for batch sizes (I wanted to do it anyhow), start with a warning first and we can switch between warning and returning an error by a configuration flag.

Sounds useful for batch operations.

**Comment by swasik** [comment]
IMO, adding warning will just delay the problem - maybe I am wrong but most probably nobody will notice it if it is hidden in the server logs and in the end people will be surprised as well. 

@mykaul if you really would like to go with warning version, let us not introduce new flag. Just let us put artificially high default limit (e.g. 2k), add metric + warning and in the next release we will have data to decide how much we can decrease the limit. But let us make decision fast.

**Comment by amnonh** [comment]
@swasik I'm not sure, in the cloud, we can install it immediately, make sure it's not going to hurt any customer, and than enable it.

**Comment by radoslawcybulski** [comment]
@amnonh 

1. warning aren't that useful. In general people learn to ignore them, unless there's a correlation between a warning and something useful (like a warning indicates loss of performance), in which case it should be an error probably.
2. histogram is a cool idea. There're structures that allow cheap approximation of histograms with low memory usage. Can we make a new task for this? Also do we really need it?

**Comment by mykaul** [comment]
> [@amnonh](https://github.com/amnonh)
> 
> 1. warning aren't that useful. In general people learn to ignore them, unless there's a correlation between a warning and something useful (like a warning indicates loss of performance), in which case it should be an error probably.

They help us diagnose issue. It's not for the users.

> 3. histogram is a cool idea. There're structures that allow cheap approximation of histograms with low memory usage. Can we make a new task for this? Also do we really need it?

**Comment by nyh** [comment]
> I think that this discussion is taking too long (at we are already >50 comments on PR for 10-liner). 

Let's be careful not to throw around that slogan "50 comments is too much" (that Dor said in his presentation this morning). Sometimes a long discussion is justified. Sometimes the developer working on an issue doesn't yet have the experience and needs inputs from other developers and managers - and inputs from several people necessarily leads to conflicting opinions and longer discussions. You can solve this problem by only letting highly experienced people work on issues and solving the conflicts alone inside their heads - but don't want to do that. You can also solve this problem by creating a detailed an **official** (can't be changed or argued after written!)  requirements document for each issue - I don't think you want that either. Dor's suggestion to "have a call instead of a 50 comment discussion" usually won't solve the problem either: Unless we're lucky, people will not have all the information during the call, and will continue to raise issues after the call ended and need more calls.

> Taking into account that average is low, I vote for merging. If some customer hits the limit they will have to change the config, we should just make sure that we are clear about it in release notes.

This is not the place to vote for merging - it is https://github.com/scylladb/scylladb/pull/23232.
That PR is currently waiting for review of its documentation - not about the "100" default.

I still think that all this discussion was not needed in the first place - we should have either had NO limit at all (and limit the implementation's concurrency) or choose DynamoDB's limit (25) as the default. But other people thought different limits should be set (1000 and 100) which caused all these comments. I'm just fine with 100 as a default.

**Comment by nyh** [comment]
> [@mykaul](https://github.com/mykaul) Ok, I misunderstood the idea then. But that is great. So [@radoslawcybulski](https://github.com/radoslawcybulski) let us add this warning (making sure that it is not printed more often than every 5 seconds - [@amnonh](https://github.com/amnonh) do you know if we have such option in our logging infrastructure?) and close the issue.

Please explain the point of a log warning. An oversized batch will cause an **error**, which the application will surely notice. Why would a logged warning help? I think it's a bad idea because we have hundreds of other API errors for which we return an API error and don't log anything, so why start with this one?

**Comment by radoslawcybulski** [comment]
@ScyllaPiotr @swasik @nyh @mykaul  @amnonh

let's move on => in which direction. Currently i'm absolutely confused, what is going on. Should i or should i not implement warning thing? With or without hard error? What should be a hard error limit if any?

Given how we can't seem to decide how it's best to proceed and given how a lot of reviews are essentially "let's do it from scratch in a different way", maybe we should instead bring the task back to the "design phase", where people more knowledgeable than me (ie you guys...) would decide, what is the best course of action from the perspective of users of our dbase? The task seems to touch more complicated issues, than i originally thought.

**Comment by ScyllaPiotr** [comment]
Wise words.
@swasik @nyh @radoslawcybulski I propose to meet and to decide collaboratively.

**Comment by nyh** [comment]
> [@ScyllaPiotr](https://github.com/ScyllaPiotr) [@swasik](https://github.com/swasik) [@nyh](https://github.com/nyh) [@mykaul](https://github.com/mykaul) [@amnonh](https://github.com/amnonh)
> 
> let's move on => in which direction. Currently i'm absolutely confused, what is going on. Should i or should i not implement warning thing?

Following Dor's advice, let's set up a call and discuss this until everyone is convinced.

Following Kant, I'm **against** adding a warning in this case because I wouldn't want to make it into a general rule - I wouldn't want every failed API request to end up on the warning log. I don't think it's useful and it doesn't make sense. Clients can see their requests are failing.

> Given how we can't seem to decide how it's best to proceed and given how a lot of reviews are essentially "let's do it from scratch in a different way",

I don't understand why we're discussing this here and not in the pull request. More importantly, two reviewers already gave your PR an approval and the last review comment was just about the documentation (which you handled as well), so I don't understand why you concluded that you were asked to "do it from scratch in a different way".

I think the direction with the configurable limit of 100 is just fine, and we don't need to redo it from scratch.
